game/glutil.cc

```diff
@@ -40,13 +52,52 @@ void VertexArray::draw(GLint mode) {
 		glEnableVertexAttribArray(vertColor);
 		glVertexAttribPointer(vertColor, 4, GL_FLOAT, GL_FALSE, stride, ptr);
 	}
+	glerror.check("enable arrays");
 	glDrawArrays(mode, 0, size());
 
 	if (vertPos != -1) glDisableVertexAttribArray(vertPos);
 	if (vertTexCoord != -1) glDisableVertexAttribArray(vertTexCoord);
 	if (vertNormal != -1) glDisableVertexAttribArray(vertNormal);
 	if (vertColor != -1) glDisableVertexAttribArray(vertColor);
 	if (m_vbo) glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
 }
 
+GLErrorChecker::GLErrorChecker(std::string const& info): info(info) {
+	stack.push_back(std::string());
+	check("before starting");
+	stack.back() = info;
+}
+
+GLErrorChecker::~GLErrorChecker() { check("after finishing"); stack.pop_back(); }
+
+void GLErrorChecker::check(std::string const& what) {
+	GLenum err = glGetError();
+	if (err != GL_NO_ERROR) {
+		stack.back() = info + " " + what;
+		// Prefix with all currently active GLErrorChecker contexts
+		std::string logmsg = "opengl/error: ";
+		for (auto s: stack) logmsg += s + ": ";
+		logmsg += msg(err) + "\n";
+		std::clog << logmsg << std::flush;
+	}
+	stack.back() = info + " after " + what;
+}
+
+/* static */ std::string GLErrorChecker::msg(GLenum err) {
+	switch(err) {
+		case GL_NO_ERROR: return std::string();
+		case GL_INVALID_ENUM: return "Invalid enum";
+		case GL_INVALID_VALUE: return "Invalid value";
+		case GL_INVALID_OPERATION: return "Invalid operation";
+		case GL_INVALID_FRAMEBUFFER_OPERATION: return "FBO is not complete";
+		case GL_STACK_OVERFLOW: return "Stack overflow";
+		case GL_STACK_UNDERFLOW: return "Stack underflow";
+		case GL_OUT_OF_MEMORY: return "Out of memory";
+		default: return "Unknown error";
+	}
+}
+
+/* static */ thread_local std::vector<std::string> GLErrorChecker::stack;
```

Comment:
Remove the static keyword if it's not needed

Reply:
This is a definition for a static member variable. Due to C++ syntax, the static keyword can only be used inside class {} and not again when the variable is defined. The comment is there to make it clear to the reader that this is in fact a static member.

Reply:
alright Granted 👍